lib/Dialect/FIRRTL/Transforms/LegacyWiring.cpp

```diff
@@ -0,0 +1,120 @@
+    } else if (auto ext = dyn_cast<FExtModuleOp>(portTarget.getOp())) {
+      if (target.instances.empty()) {
+        auto paths = state.instancePathCache.getAbsolutePaths(ext);
+        if (paths.size() > 1) {
+          mlir::emitError(state.circuit.getLoc())
+              << "cannot resolve a unique instance path from the "
+                 "external module target"
+              << target.ref;
+          return failure();
+        }
+        auto inst = cast<InstanceOp>(paths[0].back());
+        builder.setInsertionPointAfter(inst);
+        targetsValues.push_back(getValueByFieldID(
+            builder, inst->getResult(portNum), target.fieldIdx));
+      } else {
+        auto inst = cast<InstanceOp>(target.instances.back());
+        builder.setInsertionPointAfter(inst);
+        targetsValues.push_back(getValueByFieldID(
+            builder, inst->getResult(portNum), target.fieldIdx));
+      }
+    } else {
+      return mlir::emitError(state.circuit.getLoc())
+             << "Annotation has invalid target: " << anno;
+    }
+  } else if (auto opResult = target.ref.dyn_cast<OpAnnoTarget>()) {
+    if (target.isOpOfType<WireOp, RegOp, RegResetOp>()) {
+      auto module = cast<FModuleOp>(opResult.getModule());
+      builder.setInsertionPointToEnd(module.getBodyBlock());
+      if (opResult.getOp()->getNumResults() != 1) {
+        // Unsure about what ops emit more/less than 1 result, perhaps the type
+        // check is enough
+        return failure();
+      }
+      targetsValues.push_back(getValueByFieldID(
+          builder, opResult.getOp()->getResult(0), target.fieldIdx));
+    } else {
+      return mlir::emitError(state.circuit.getLoc())
+             << "Annotation targets non-wireable operation: " << anno;
+    }
+  } else {
+    return mlir::emitError(state.circuit.getLoc())
+           << "Annotation has invalid target: " << anno;
+  }
+
+  // Get pin field
+  auto pin = anno.getAs<StringAttr>("pin");
+  if (!pin) {
+    return mlir::emitError(state.circuit.getLoc())
+           << "Annotation does not have an associated pin name: " << anno;
+  }
+
+  bool problemAlreadyDefined =
+      state.legacyWiringProblems.find(pin) != state.legacyWiringProblems.end();
+
+  // Handle difference between sinks and sources
+  auto clazz = anno.getAs<StringAttr>("class").getValue();
+  if (clazz == wiringSourceAnnoClass) {
+    if (problemAlreadyDefined) {
+      // Attempt to update the existing problem
+      if (state.legacyWiringProblems[pin].source || targetsValues.size() != 1) {
+        return mlir::emitError(state.circuit.getLoc())
+               << "More than one " << wiringSourceAnnoClass
+               << " defined for pin " << pin;
+      }
+      state.legacyWiringProblems[pin].source = targetsValues[0];
+    } else {
+      // Create a new problem
+      state.legacyWiringProblems[pin] = {targetsValues[0]};
+    }
+  } else if (clazz == wiringSinkAnnoClass) {
+    if (problemAlreadyDefined) {
```

Comment:
Could this if/else be reduced to
`llvm::append_range(state.legacyWiringProblems[pin], targetsValues)` ?
(let the default constructor create the empty problem in the else case?)

If not or don't like that, might be a more readable alternative to  `X.insert(X.end, Y.begin(), Y.end())` :+1: .

Reply:
I don't even know why I made targetsValues a Vector :sweat_smile:, target always maps to one Value here and the code is a lot nicer if I just use defaults. e.g.
```
  else if (clazz == wiringSinkAnnoClass) {
    state.legacyWiringProblems[pin].sinks.push_back(targetValue);
  }
  ````